File: src/main/java/org/lovaprod/service/WalletService.java
package org.lovaprod.service;

import org.lovaprod.model.Transaction;
import org.lovaprod.model.User;
import org.lovaprod.utils.MoneyUtils;

import java.util.List;
import java.util.Scanner;

public class WalletService {


    public static void checkBalance(User user) {
        // Display user's wallet balance
        System.out.println("Wallet Balance: " + user.getWallet().getBalance() + MoneyUtils.MONEY_UNIT);
    }

    public static void addFunds(Scanner scanner, User user) {
        // Allow user to add funds to the wallet
        System.out.print("Enter the amount (in"+MoneyUtils.MONEY_UNIT+") to add: ");
        double amount = scanner.nextDouble();
        scanner.nextLine();

        if (MoneyUtils.isAmountValid(amount)){
            user.getWallet().deposit(amount);
            System.out.println("Funds added successfully. New balance: " + user.getWallet().getBalance() + MoneyUtils.MONEY_UNIT);
        }
    }

    public static void withdrawMoney(Scanner scanner, User user) {
        // Allow user to withdraw money from the wallet
        System.out.print("Enter the amount (in"+MoneyUtils.MONEY_UNIT+") to withdraw: ");
        double amount = scanner.nextDouble();
        scanner.nextLine(); // Consume the newline

        if (user.getWallet().withdraw(amount)) {
            System.out.println("Withdrawal successful. New balance: " + user.getWallet().getBalance() + MoneyUtils.MONEY_UNIT);
        }
    }

    public static void peerToPeerTransfers(Scanner scanner, User user, List<User> users) {
        // Allow user to make peer-to-peer transfers
        System.out.print("Enter the destination username: ");
        String destinationUsername = scanner.nextLine();

        User destinationUser = AccountService.findUserByUsername(destinationUsername, users);

        if (destinationUser != null) {
            System.out.print("Enter the amount (in"+MoneyUtils.MONEY_UNIT+") to transfer: ");
            double amount = scanner.nextDouble();
            scanner.nextLine(); // Consume the newline

            if (user.getWallet().withdraw(amount, destinationUser)) {
                System.out.println(amount + MoneyUtils.MONEY_UNIT + " send to : " + destinationUsername);
                System.out.println("Transfer successful. New balance: " + user.getWallet().getBalance() + MoneyUtils.MONEY_UNIT);
            }
        } else {
            System.out.println("User not found with the provided username.");
        }
    }

    public static void transactionHistory(User user) {
        // Display transaction history
        List<Transaction> transactions = user.getWallet().getTransactionHistory().getTransactions();

        System.out.println("Transaction History:");
        for (Transaction transaction : transactions) {
            System.out.println("Transaction ID: " + transaction.getId());
            System.out.println("Amount: " + transaction.getAmount() + MoneyUtils.MONEY_UNIT);
            System.out.println("Type: " + transaction.getType());
            System.out.println("Destination: " + transaction.getDestination());
            System.out.println("Transaction Date: " + transaction.getTransactionDate());
            System.out.println("------");
        }
    }
}

File: WalletService.cs
using System;
using System.Collections.Generic;

namespace WalletApp
{
    public class WalletService
    {
        public static void CheckBalance(User user)
        {
            Console.WriteLine("Wallet Balance: " + user.Wallet.Balance + MoneyUtils.MoneyUnit);
        }

        public static void AddFunds(User user)
        {
            Console.Write("Enter the amount (in" + MoneyUtils.MoneyUnit + ") to add: ");
            double amount = Convert.ToDouble(Console.ReadLine());

            if (MoneyUtils.IsAmountValid(amount))
            {
                user.Wallet.Deposit(amount);
                Console.WriteLine("Funds added successfully. New balance: " + user.Wallet.Balance + MoneyUtils.MoneyUnit);
            }
        }

        public static void WithdrawMoney(User user)
        {
            Console.Write("Enter the amount (in" + MoneyUtils.MoneyUnit + ") to withdraw: ");
            double amount = Convert.ToDouble(Console.ReadLine());

            if (user.Wallet.Withdraw(amount))
            {
                Console.WriteLine("Withdrawal successful. New balance: " + user.Wallet.Balance + MoneyUtils.MoneyUnit);
            }
        }

        public static void PeerToPeerTransfers(User user, List<User> users)
        {
            Console.Write("Enter the destination username: ");
            string destinationUsername = Console.ReadLine();

            User destinationUser = AccountService.FindUserByUsername(destinationUsername, users);

            if (destinationUser != null)
            {
                Console.Write("Enter the amount (in" + MoneyUtils.MoneyUnit + ") to transfer: ");
                double amount = Convert.ToDouble(Console.ReadLine());

                if (user.Wallet.Withdraw(amount, destinationUser))
                {
                    Console.WriteLine(amount + MoneyUtils.MoneyUnit + " sent to : " + destinationUsername);
                    Console.WriteLine("Transfer successful. New balance: " + user.Wallet.Balance + MoneyUtils.MoneyUnit);
                }
            }
            else
            {
                Console.WriteLine("User not found with the provided username.");
            }
        }

        public static void TransactionHistory(User user)
        {
            List<Transaction> transactions = user.Wallet.TransactionHistory.Transactions;

            Console.WriteLine("Transaction History:");
            foreach (Transaction transaction in transactions)
            {
                Console.WriteLine("Transaction ID: " + transaction.Id);
                Console.WriteLine("Amount: " + transaction.Amount + MoneyUtils.MoneyUnit);
                Console.WriteLine("Type: " + transaction.Type);
                Console.WriteLine("Destination: " + transaction.Destination);
                Console.WriteLine("Transaction Date: " + transaction.TransactionDate);
                Console.WriteLine("------");
            }
        }
    }
}
